src/time_tracker/invoices/core/spec.clj

```diff
@@ -87,40 +88,73 @@
   (= (set (:required-user-ids args))
      (set (keys ret))))
 
-(s/fdef build-user-id->hours
+(s/fdef invoices-core/build-user-id->hours
         :args (s/with-gen (s/cat :required-user-ids (s/coll-of ::core-spec/id)
                                  :timers            ::timers)
                 build-user-id->hours-args-gen)
-        :ret ::user-id->hours)
-
-(s/def ::id->name-map
-  (s/map-of ::core-spec/id string?))
+        :ret ::user-id->hours
+        :fn build-user-id->hours-pred)
 
-(defn- csv-rows-ret
-  [ret]
-  (let [names        (mapv first ret)
-        unique-names (set names)]
-    ;; The names must be unique.
-    (= (count names) (count unique-names))))
+(defn- user-id->rate-gen [users]
+  (gen/bind (gen/vector (s/gen ::invoices-spec/rate) (count users))
+            (fn [rates]
+              (gen/return (zipmap (keys users) rates)))))
 
-(defn- csv-rows-args-gen []
+(defn- user-amounts-args-gen []
   (gen/bind (users-timers-gen)
             (fn [[users timers]]
               (gen/tuple
-               (gen/return (invoices-core/build-user-id->hours (keys users) timers))
-               (gen/return (invoices-core/id->name users))))))
-
-(s/fdef invoices-core/csv-rows
-        :args (s/with-gen (s/cat :user-id->hours ::user-id->hours
-                                 :user-id->name ::id->name-map)
-                csv-rows-args-gen)
-        :ret (s/and (s/coll-of
-                     (s/cat :user-name string?
-                            :hours ::core-spec/positive-num))
-                    csv-rows-ret))
-
-(s/fdef invoices-core/generate-csv
+                (gen/return users)
+                (gen/return timers)
+                (user-id->rate-gen users)))))
+
+(defn- user-amounts-pred
+  [{:keys [args ret]}]
+  (let [unique-args-ids (->> (:users args)
+                             (keys)
+                             (set))
+        unique-ret-ids (set (map :id ret))]
+    ;; Every user passed should be present in the returned rows.
+    (= unique-args-ids unique-ret-ids)))
+
+(s/fdef invoices-core/user-hours
         :args (s/with-gen (s/cat :users ::users
-                                 :timers ::timers)
-                users-timers-gen)
-        :ret (s/spec string?))
+                                 :timers ::timers
+                                 :user-id->rate ::invoices-spec/user-id->rate)
+                          user-amounts-args-gen)
+        :ret ::invoices-spec/user-hours
+        :fn user-amounts-pred)
+
+(s/def ::user-id->name (s/fspec :args (s/cat :id ::users-spec/id)
+                                :ret ::users-spec/name))
+
+(s/fdef invoices-core/invoice-items
+        :args (s/cat :invoice ::invoices-spec/invoice
+                     :user-id->name ::user-id->name)
+        :ret ::invoices-spec/items)
+
+(s/fdef invoices-core/subtotal
+        :args (s/cat :invoice ::invoices-spec/invoice)
+        :ret ::core-spec/money-val)
```

Comment:
Nothing is testing the logic here. This may be a good place to use some example based tests to check behaviour.